engines/alg/game_crimepatrol.cpp

```diff
@@ -0,0 +1,1352 @@
+	if (it != _rectHitFuncs.end()) {
+		return (*(*it)._value);
+	} else {
+		error("Could not find rectHit function: %s", name.c_str());
+	}
+}
+
+CPScriptFunctionScene GameCrimePatrol::getScriptFunctionScene(SceneFuncType type, Common::String name) {
+	CPScriptFunctionSceneMap *functionMap;
+	switch (type) {
+	case PREOP:
+		functionMap = &_scenePreOps;
+		break;
+	case SHOWMSG:
+		functionMap = &_sceneShowMsg;
+		break;
+	case INSOP:
+		functionMap = &_sceneInsOps;
+		break;
+	case WEPDWN:
+		functionMap = &_sceneWepDwn;
+		break;
+	case SCNSCR:
+		functionMap = &_sceneScnScr;
+		break;
+	case NXTFRM:
+		functionMap = &_sceneNxtFrm;
+		break;
+	case NXTSCN:
+		functionMap = &_sceneNxtScn;
+		break;
+	default:
+		error("Unkown scene script type: %u", type);
+		break;
+	}
+	CPScriptFunctionSceneMap::iterator it;
+	it = functionMap->find(name);
+	if (it != functionMap->end()) {
+		return (*(*it)._value);
+	} else {
+		error("Could not find scene type %u function: %s", type, name.c_str());
+	}
+}
+
+void GameCrimePatrol::callScriptFunctionRectHit(Common::String name, Rect *rect) {
+	CPScriptFunctionRect function = getScriptFunctionRectHit(name);
+	function(rect);
+}
+
+void GameCrimePatrol::callScriptFunctionScene(SceneFuncType type, Common::String name, Scene *scene) {
+	CPScriptFunctionScene function = getScriptFunctionScene(type, name);
+	function(scene);
+}
+
+Common::Error GameCrimePatrol::run() {
+	init();
+	_NewGame();
+	_cur_scene = _startscene;
+	Common::String oldscene;
+	while (!_vm->shouldQuit()) {
+		oldscene = _cur_scene;
+		_SetFrame();
+		_fired = false;
+		Scene *scene = _sceneInfo->findScene(_cur_scene);
+		if (!loadScene(scene)) {
+			if (scene->nxtscn == "CONTINUE_TARGET_PRACTICE") {
+				callScriptFunctionScene(NXTSCN, scene->nxtscn, scene);
+			} else {
+				error("Cannot find scene %s in libfile", scene->name.c_str());
+			}
+		}
+		_sceneSkipped = false;
+		Audio::PacketizedAudioStream *audioStream = _videoDecoder->getAudioStream();
+		g_system->getMixer()->stopHandle(_sceneAudioHandle);
+		g_system->getMixer()->playStream(Audio::Mixer::kPlainSoundType, &_sceneAudioHandle, audioStream, -1, Audio::Mixer::kMaxChannelVolume, 0, DisposeAfterUse::NO);
+		_paletteDirty = true;
+		_nextFrameTime = _GetMsTime() + 100;
+		callScriptFunctionScene(PREOP, scene->preop, scene);
+		_currentFrame = _GetFrame(scene);
+		while (_currentFrame <= scene->endFrame && _cur_scene == oldscene && !_vm->shouldQuit()) {
+			_UpdateMouse();
+			callScriptFunctionScene(SHOWMSG, scene->scnmsg, scene);
+			callScriptFunctionScene(INSOP, scene->insop, scene);
+			_holster = _WeaponDown();
+			if (_holster) {
+				callScriptFunctionScene(WEPDWN, scene->wepdwn, scene);
+			}
+			Common::Point firedCoords;
+			if (__Fired(&firedCoords)) {
+				if (!_holster) {
+					Rect *hitGlobalRect = _CheckZone(_menuzone, &firedCoords);
+					if (hitGlobalRect != nullptr) {
+						callScriptFunctionRectHit(hitGlobalRect->rectHit, hitGlobalRect);
+					} else {
```

Comment:
```suggestion
					} else if (_shots > 0) {
```
This will eliminate one level of indentation

Reply:
fixed